Move all id/name functions to SedBase.cpp

luciansmith opened the pull request:
This change moves all functionality with respect to id and name to SedBase, so the bulk of the change is simply 'deleted code'.

In L1v4, everything has an id and name, so everything works fine there.

In L1v3 and earlier, some elements had id and name, some had just id, some had just name, and some had neither.  This is now handled with the 'mIdAllowedPreV4' and 'mNameAllowedPreV4' boolean member variables on SedBase, which the relevant classes define in their constructors, and then checked in the various SedBase functions that set/get/unset/read/write the id and name attributes.  The exception is that the 'id is required' functionality is still handled in 'readAttributes', but instead of setting the id and then checking if it's set, it now just calls SedBase::readAttributes and then checks 'isSetId()'.

luciansmith committed:
Move all id/name functions to SedBase.cpp

This change moves all functionality with respect to id and name to SedBase, so the bulk of the change is simply 'deleted code'.

In L1v4, everything has an id and name, so everything works fine there.

In L1v3 and earlier, some elements had id and name, some had just id, some had just name, and some had neither.  This is now handled with the 'mIdAllowedPreV4' and 'mNameAllowedPreV4' boolean member variables on SedBase, which the relevant classes define in their constructors, and then checked in the various SedBase functions that set/get/unset/read/write the id and name attributes.  The exception is that the 'id is required' functionality is still handled in 'readAttributes', but instead of setting the id and then checking if it's set, it now just calls SedBase::readAttributes and then checks 'isSetId()'.

fbergmann commented on src/sedml/SedAbstractCurve.cpp:
```diff
@@ -68,6 +67,7 @@ SedAbstractCurve::SedAbstractCurve(unsigned int level, unsigned int version)
   , mElementName("abstractCurve")
 {
   setSedNamespacesAndOwn(new SedNamespaces(level, version));
+  mNameAllowedPreV4 = true;
```
should we not have id allowed prev4 too?

luciansmith replied:
Ah, this seems to be a UML vs. the text thing.  The UML diagram doesn't have it, and the text doesn't list it, but it's in the table.  (This was true in l1v2, too).

For me, if the spec was ambiguous, I'm OK with allowing the ID in those level/versions, especially since we added it later.  I'll make the fix.

luciansmith replied:
...except now that I look at the other tables, every single table in the spec that lists 'id' also lists 'name', and visa versa.  So while the UML diagrams indicated that some classes only had 'name' and some only had 'id', the tables always listed them together.

Do we just say that everything had both 'id' and 'name' that had either?  I suppose this might be a question for the editors.  My inclination is to allow both.

luciansmith replied:
OK, here's a breakdown of the only-id-or-name classes:

* Range:  only 'id' in the UML; no table with allowed attributes, so nothing ever mentions 'name' being allowed.
* AbstractCurve:  only 'name' in the UML; both 'id' and 'name' in Table 2.28
* DataSet:  only 'name' in the UML; both 'id' and 'name' in Table 2.30
* Surface:  only 'name' in the UML; both 'id' and 'name' in Table 2.29  (Also, inherits from Curve.)

I've filed https://github.com/SED-ML/sed-ml/issues/139 to ask what other editors think.

fbergmann replied:
well, since libSEDML always allowed to set an id attribute on these elements, i'm hesitant to just remove that for now, so i would want it to be set / read / written from now on too (if the user would set it)

fbergmann commented on src/sedml/SedBase.cpp:
```diff
@@ -281,13 +293,20 @@ SedBase::getMetaId ()
 {
-  static string empty;
-  return empty;
+    if (!mNameAllowedPreV4 && getVersion() < 4 && getLevel() == 1) {
```
is this really necessary? I'd let people return the member variable in any case.

luciansmith replied:
In theory, I'd be OK with letting people get it?  The issue is more that 'getName' (and 'getId') is used internally a lot, and I'm hesitant to tell that code that an attribute is set if it's not going to be written out.  (For instance, 'isSetName' calls 'getName' and checks if that's empty.)

If people need to get the name, they can always just set the level to 4, and then 'getName' will work if there's something there.

fbergmann commented on src/sedml/SedBase.cpp:
```diff
@@ -1678,7 +1708,16 @@ SedBase::unsetId ()
 int SedBase::unsetName()
 {
-  return LIBSEDML_OPERATION_FAILED;
+    mName.erase();
```
i dont see how this could ever fail

luciansmith replied:
Neither do I, but you see it in a lot of libsbml code, and it's generated by Deviser, where it landed in some other function I copied and pasted into here.  I'll go ahead and remove the fail condition; I also don't think it's possible to reach it.

fbergmann commented on src/sedml/SedFitExperiment.cpp:
```diff
@@ -144,16 +144,6 @@ SedFitExperiment::~SedFitExperiment()
 }
 
```
you are not changing the id / name allowed members here, cause it is l1v4? Would that ever cause an issue?

luciansmith replied:
Right; I assumed that anything that was added in l1v4 is already handled and not written out in conversions to l1v3.  (If not, I guess that's a separate issue we should file?)

fbergmann replied:
ok

fbergmann commented:
@luciansmith if you could just double check on the id allowed on curve, then i would be happy to merge this back. thanks for doing this.

luciansmith committed:
Clean up code that can't be reached.

fbergmann committed:
- allow ids to be set / read from curve/dataset/surface

libsedml always wrote these ids, and even required them to be set at some point, it would be strange to now not do so

fbergmann committed:
Merge branch 'level1-version4' into issue-84
